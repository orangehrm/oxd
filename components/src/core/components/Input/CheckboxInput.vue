<template>
  <div class="oxd-checkbox-wrapper">
    <label :class="{'--disabled': disabled}">
      <template v-if="labelPosition === 'left'">
        {{ optionLabel }}
      </template>
      <input
        type="checkbox"
        @focus="onFocus"
        @blur="onBlur"
        @click="onClick"
        @change="onChange"
        :disabled="disabled"
        v-bind="$attrs"
        v-model="checked"
      />
      <span :class="classes" :style="style" class="oxd-checkbox-input">
        <oxd-icon class="oxd-checkbox-input-icon" :name="checkIcon" />
      </span>
      <template v-if="labelPosition === 'right'">
        {{ optionLabel }}
      </template>
    </label>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {Position, LABEL_POSITIONS, RIGHT} from './types';
import Icon from '@orangehrm/oxd/core/components/Icon/Icon.vue';

export interface State {
  focused: boolean;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  checkedProxy: any;
}

export default defineComponent({
  name: 'oxd-checkbox-input',
  inheritAttrs: false,
  props: {
    modelValue: {},
    style: {
      type: Object,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    checkIcon: {
      type: String,
      default: 'check',
    },
    optionLabel: {
      type: String,
      default: '',
    },
    labelPosition: {
      type: String,
      default: RIGHT,
      validator: function(value: Position) {
        return LABEL_POSITIONS.indexOf(value) !== -1;
      },
    },
  },

  emits: ['update:modelValue', 'blur', 'focus', 'change'],

  components: {
    'oxd-icon': Icon,
  },

  data(): State {
    return {
      focused: false,
      checkedProxy: {},
    };
  },

  computed: {
    classes(): object {
      return {
        'oxd-checkbox-input': true,
        'oxd-checkbox-input--active': !this.focused,
        'oxd-checkbox-input--focus': this.focused,
        'oxd-checkbox-input--error': this.hasError,
        [`--label-${this.labelPosition}`]: true,
      };
    },
    checked: {
      get() {
        return this.modelValue;
      },
      set(value) {
        this.checkedProxy = value;
      },
    },
  },

  methods: {
    onFocus(e: Event) {
      this.focused = true;
      this.$emit('focus', e);
    },
    onBlur(e: Event) {
      this.focused = false;
      this.$emit('blur', e);
    },
    onClick(e: Event) {
      this.focused = false;
      this.$emit('click', e);    
    },
    onChange(e: Event) {
      this.$emit('update:modelValue', this.checkedProxy);
      this.$emit('change', e);
    },
  },
});
</script>

<style src="./checkbox-input.scss" lang="scss" scoped></style>
