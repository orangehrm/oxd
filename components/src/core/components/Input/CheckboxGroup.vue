<script lang="ts">
import {defineComponent, h} from 'vue';
import CheckboxInput from '@orangehrm/oxd/core/components/Input/CheckboxInput.vue';
import InputGroup from '@orangehrm/oxd/core/components/InputField/InputGroup.vue';
import useTranslate from '../../../composables/useTranslate';

export interface State {
  focused: boolean;
}

export interface Options {
  id: number;
  label: string;
  disabled?: boolean;
  style?: Record<string, any>;
}

export default defineComponent({
  name: 'oxd-checkbox-group',
  components: {
    CheckboxInput,
    InputGroup,
  },

  emits: ['update:modelValue', 'blur', 'focus', 'change'],

  props: {
    id: {
      type: String,
      default: 'check-box-group-id',
    },
    class: {
      type: String,
      default: 'check-box-column',
    },
    style: {
      type: Object,
    },
    disabled: {
      type: Boolean,
    },
    options: {
      type: Array,
    },
    modelValue: {
      type: Array,
      default: () => [],
    },
  },

  render() {
    const inputId = this.id == '' ? 'check-box-group-id' : this.id;
    const inputClass = this.class == '' ? 'check-box-column' : this.class;
    const {$t} = useTranslate();
    return h(
      InputGroup,
      {
        classes: {
          wrapper: inputClass,
        },
        id: inputId,
        style: this.style,
      },
      [
        this.options.map((option: Options, index: number) => {
          return h(CheckboxInput, {
            id: inputId + '_' + option.id,
            value: option.id,
            style: option.style,
            autofocus: this.$attrs.autofocus && index === 0 ? true : false,
            optionLabel: $t(option.label),
            modelValue: this.modelValue,
            disabled: this.disabled ? 'true' : option.disabled,
            onFocus: () => {
              this.$emit('focus', true);
            },
            onBlur: () => {
              this.$emit('blur', true);
            },
            onClick: () => {
              this.$emit('click', true);
            },
            onChange: () => {
              let idArray: number[];
              if (this.modelValue.includes(option.id)) {
                idArray = [...this.modelValue].filter(
                  value => value !== option.id,
                );
              } else {
                idArray = [...this.modelValue, option.id];
              }
              this.$emit('update:modelValue', idArray.sort());
              this.$emit('change', true);
            },
          });
        }),
      ],
    );
  },
});
</script>

<style lang="scss" scoped>
:deep(.check-box-row) {
  display: flex;
  gap: 2rem;
}

:deep(.check-box-column) {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}
</style>
