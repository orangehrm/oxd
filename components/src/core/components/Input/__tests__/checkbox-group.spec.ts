import {mount} from '@vue/test-utils';
import CheckboxGroup from '@orangehrm/oxd/core/components/Input/CheckboxGroup.vue';
import CheckboxInput from '@orangehrm/oxd/core/components/Input/CheckboxInput.vue';

const optionsList = [
  {
    id: 1,
    label: 'Item one',
  },
  {
    id: 2,
    label: 'Item two',
  },
];

describe('CheckboxGroup > CheckboxGroup.vue', () => {
  it('renders OXD CheckboxGroup', () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
        class: 'check-box-column',
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should accept valid input', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#check-box-group-id_1').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toEqual([[[1]]]);
  });

  it('should emit checked value on click', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#check-box-group-id_1').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toEqual([[[1]]]);
  });

  it('should return both the selected values', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#check-box-group-id_1').trigger('click');
    wrapper.find('#check-box-group-id_2').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toEqual([[[1]], [[2]]]);
  });

  it('should uncheck the previous selected values', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
        modelValue: [1],
      },
    });
    wrapper.find('#check-box-group-id_1').trigger('change');
    expect(wrapper.emitted('update:modelValue')).toEqual([[[]]]);
  });

  it('should focus the checkbox', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#check-box-group-id_1').trigger('focus');
    expect(wrapper.emitted('focus')).toEqual([[true]]);
  });

  it('should blur the checkbox', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
      },
    });
    wrapper.find('#check-box-group-id_1').trigger('blur');
    expect(wrapper.emitted('blur')).toEqual([[true]]);
  });

  it('should change the class and id', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
        id: 'check-box-new-id',
        class: 'check-box-new-class',
      },
    });

    expect(wrapper.find('.check-box-new-class').exists()).toBeTruthy();
    expect(wrapper.find('#check-box-new-id_1').exists()).toBeTruthy();
  });

  it('should return the default the class and id', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
        id: '',
        class: '',
      },
    });

    expect(wrapper.find('.check-box-column').exists()).toBeTruthy();
    expect(wrapper.find('#check-box-group-id_1').exists()).toBeTruthy();
  });

  it('should not emit any value if disabled', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
        disabled: true,
      },
    });
    wrapper.find('#check-box-group-id_1').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted('update:modelValue')).toBeFalsy();
  });

  it('autofocus should apply to first element of checkboxgroup', async () => {
    const wrapper = mount(CheckboxGroup, {
      props: {
        options: optionsList,
        autofocus: true,
      },
    });
    const firstCheckBox = wrapper
      .findAllComponents(CheckboxInput)[0]
      .find('input[type=checkbox]');
    const secondCheckBox = wrapper
      .findAllComponents(CheckboxInput)[1]
      .find('input[type=checkbox]');
    expect(firstCheckBox.attributes('autofocus')).toEqual('');
    expect(secondCheckBox.attributes('autofocus')).toEqual(undefined);
  });
});
