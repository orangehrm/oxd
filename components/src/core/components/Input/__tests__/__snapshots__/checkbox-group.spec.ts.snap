// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup > CheckboxGroup.vue renders OXD CheckboxGroup 1`] = `
<div class="oxd-input-group">
  <div class="oxd-input-group__label-wrapper">
    <!--v-if-->
    <!--v-if-->
    <!--v-if-->
  </div>
  <div class="check-box-column">
    <div class="oxd-checkbox-wrapper"><label class="">
        <!--v-if--><input type="checkbox" true-value="true" false-value="false" id="check-box-group-id_1" value="1"><span class="oxd-checkbox-input oxd-checkbox-input--active --label-right oxd-checkbox-input"><i class="oxd-svg-icon oxd-svg-icon--xx-small oxd-checkbox-input-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 36.51 27.2"><defs><style>{fill:currentColor;}</style></defs><g id="Layer_2" data-name="Layer 2"><g id="_ÎÓÈ_1" data-name="—ÎÓÈ_1"><path style="fill:currentColor" d="M0,13c.11-.32.21-.64.34-1a4.23,4.23,0,0,1,6.94-1.28c1.77,1.74,3.52,3.51,5.27,5.26.71.71,1.22.71,1.93,0Q21.79,8.68,29.09,1.38A4.23,4.23,0,0,1,32.22,0,4,4,0,0,1,36.1,2.5,4.08,4.08,0,0,1,35.44,7c-.08.1-.17.19-.26.29L16.59,25.89a4.22,4.22,0,0,1-3.23,1.31,4.1,4.1,0,0,1-2.85-1.26Q5.88,21.34,1.26,16.7A4.32,4.32,0,0,1,.08,14.56c0-.07,0-.15-.08-.18Z"></path></g></g></svg></i></span>Item one
      </label></div>
    <div class="oxd-checkbox-wrapper"><label class="">
        <!--v-if--><input type="checkbox" true-value="true" false-value="false" id="check-box-group-id_2" value="2"><span class="oxd-checkbox-input oxd-checkbox-input--active --label-right oxd-checkbox-input"><i class="oxd-svg-icon oxd-svg-icon--xx-small oxd-checkbox-input-icon"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 36.51 27.2"><defs><style>{fill:currentColor;}</style></defs><g id="Layer_2" data-name="Layer 2"><g id="_ÎÓÈ_1" data-name="—ÎÓÈ_1"><path style="fill:currentColor" d="M0,13c.11-.32.21-.64.34-1a4.23,4.23,0,0,1,6.94-1.28c1.77,1.74,3.52,3.51,5.27,5.26.71.71,1.22.71,1.93,0Q21.79,8.68,29.09,1.38A4.23,4.23,0,0,1,32.22,0,4,4,0,0,1,36.1,2.5,4.08,4.08,0,0,1,35.44,7c-.08.1-.17.19-.26.29L16.59,25.89a4.22,4.22,0,0,1-3.23,1.31,4.1,4.1,0,0,1-2.85-1.26Q5.88,21.34,1.26,16.7A4.32,4.32,0,0,1,.08,14.56c0-.07,0-.15-.08-.18Z"></path></g></g></svg></i></span>Item two
      </label></div>
  </div>
  <!--v-if-->
  <!--v-if-->
</div>
`;
